### PR TITLE
feat: init of F4E_644_CBs

### PR DESCRIPTION
**Pull request by acstott**
Ran unit tests, output was: 

...................................................................................................................................................................................................................................................
Ran 243 tests in 1.975 seconds, 243 successes, 0 failures
OK

### CODE REVIEW

**Comment by charliefoxtwo** [comment]
Did you perhaps run your tests on code different from what was committed here? Identifiers like `AIL Feel-Trim CB` won't pass the tests, as all identifiers may only be a combination of upper case alphanumerics and underscores (no spaces, hyphens, or lowercase)

File: Scripts/DCS-BIOS/lib/modules/aircraft_modules/F-4E.lua [hunk 1]
```diff
@@ -958,6 +958,23 @@ local GREASE_PENCIL_DEVICE_ID = 82
...
+F_4E:defineToggleSwitch("Speed Brake CB", CIRCUIT_BREAKERS_DEVICE_ID, 3001, 1385, PILOT_BREAKER_PANEL, "Speed Brake CB")
+F_4E:defineToggleSwitch("Flaps CB", CIRCUIT_BREAKERS_DEVICE_ID, 3001, 1387, PILOT_BREAKER_PANEL, "Flaps CB")
+F_4E:defineToggleSwitch("Trim Controls CB", CIRCUIT_BREAKERS_DEVICE_ID, 3001, 1388, PILOT_BREAKER_PANEL, "Trim Controls CB")
+F_4E:defineToggleSwitch("Rudder Trim CB", CIRCUIT_BREAKERS_DEVICE_ID, 3001, 1389, PILOT_BREAKER_PANEL, "Rudder Trim CB")
+F_4E:defineToggleSwitch("Pilot Instruments CB", CIRCUIT_BREAKERS_DEVICE_ID, 3001, 2710, PILOT_BREAKER_PANEL, "Pilot Instruments Lights CB")
+F_4E:defineToggleSwitch("Pilot Warning Lights CB", CIRCUIT_BREAKERS_DEVICE_ID, 3001, 2889, PILOT_BREAKER_PANEL, "Pilot Warning Lights CB")
+F_4E:defineToggleSwitch("Landing Gear CB", CIRCUIT_BREAKERS_DEVICE_ID, 3001, 1386, PILOT_BREAKER_PANEL, "Landing Gear CB")
+
+F_4E:defineToggleSwitch("WSO Instruments CB", CIRCUIT_BREAKERS_DEVICE_ID, 3001, 2713, PILOT_BREAKER_PANEL, "WSO Instruments Lights CB")
+F_4E:defineToggleSwitch("WSO Warning Lights CB", CIRCUIT_BREAKERS_DEVICE_ID, 3001, 2890, PILOT_BREAKER_PANEL, "WSO Warning Lights CB")
```
Comment (charliefoxtwo) [inline Scripts/DCS-BIOS/lib/modules/aircraft_modules/F-4E.lua:976]: I believe these should be on the WSO CB Panel

**Comment by charliefoxtwo** [comment]
I also just noticed that all of your command values are set to 3001 - but each control within a device should have its own command.

Check out `command_defs.lua` - you should find the definitions for those constants there.

**Comment by acstott** [comment]
> PLT_CB_AIL_FEEL_TRIM

Ok, I cleaned up the PR a bit, still need to test in game before it is ready

File: Scripts/DCS-BIOS/lib/modules/aircraft_modules/F-4E.lua [hunk 2]
```diff
@@ -958,6 +958,25 @@ local GREASE_PENCIL_DEVICE_ID = 82
...
+F_4E:defineToggleSwitch("PLT_CB_SPEED_BRAKE_RW", CIRCUIT_BREAKERS_DEVICE_ID, 3005, 1385, PILOT_BREAKER_PANEL, "Speed Brake CB")
+F_4E:defineToggleSwitch("PT_CB_FLAPS_RW", CIRCUIT_BREAKERS_DEVICE_ID, 3006, 1387, PILOT_BREAKER_PANEL, "Flaps CB")
+F_4E:defineToggleSwitch("PLT_CB_TRIM_CONT_RW", CIRCUIT_BREAKERS_DEVICE_ID, 3007, 1388, PILOT_BREAKER_PANEL, "Trim Controls CB")
+F_4E:defineToggleSwitch("PLT_CB_RUD_TRIM_RW", CIRCUIT_BREAKERS_DEVICE_ID, 3008, 1389, PILOT_BREAKER_PANEL, "Rudder Trim CB")
+F_4E:defineToggleSwitch("PLT_CB_FLOODLIGHT_INSTRUMENTS", CIRCUIT_BREAKERS_DEVICE_ID, 3009, 2710, PILOT_BREAKER_PANEL, "Pilot Instruments Lights CB")
+F_4E:defineToggleSwitch("PLT_CB_INDICATOR_WARNING_LIGHTS_BRIGHTNESS", CIRCUIT_BREAKERS_DEVICE_ID, 3010, 2889, PILOT_BREAKER_PANEL, "Pilot Warning Lights CB")
+F_4E:defineToggleSwitch("PLT_CB_LANDING_GEAR_RW", CIRCUIT_BREAKERS_DEVICE_ID, 3011, 1386, PILOT_BREAKER_PANEL, "Landing Gear CB")
+
+F_4E:defineToggleSwitch("WSO_FLOODLIGHT_INSTRUMENTS", CIRCUIT_BREAKERS_DEVICE_ID, 3012, 2713, WSO_BREAKER_PANEL, "WSO Instruments Lights CB")
```
Comment (charliefoxtwo) [inline Scripts/DCS-BIOS/lib/modules/aircraft_modules/F-4E.lua:977]: Along the same naming lines, `WSO_CB_... ` would probably be good

File: Scripts/DCS-BIOS/lib/modules/aircraft_modules/F-4E.lua [hunk 3]
```diff
@@ -958,6 +958,25 @@ local GREASE_PENCIL_DEVICE_ID = 82
 -- Circuit Breakers
 local CIRCUIT_BREAKERS_DEVICE_ID = 84
 
+local PILOT_BREAKER_PANEL = "PLT CB Panel"
+
+local WSO_BREAKER_PANEL = "WSO CB Panel"
+
+F_4E:defineToggleSwitch("ARI_CIRCUIT_BREAKER", CIRCUIT_BREAKERS_DEVICE_ID, 3001, 1382, PILOT_BREAKER_PANEL, "Aileron-Rudder Interconnect (ARI) CB")
```
Comment (charliefoxtwo) [inline Scripts/DCS-BIOS/lib/modules/aircraft_modules/F-4E.lua:965]: ID should probably be `PLT_CB_ARI`

**Comment by charliefoxtwo** [comment]
I see some of the ids have LW or RW after them - what did that stand for?

**Comment by acstott** [comment]
> I see some of the ids have LW or RW after them - what did that stand for?

Left Wall / Right Wall - I was trying to retain the nomenclature from DCS as they write out like:

```PILOT_RIGHT_WALL_LANDING_GEAR_CIRCUIT_BREAKER_TOGGLE```

I can remove if it is unnecessary or confusing

**Comment by charliefoxtwo** [comment]
> > I see some of the ids have LW or RW after them - what did that stand for?
> 
> Left Wall / Right Wall - I was trying to retain the nomenclature from DCS as they write out like:
> 
> `PILOT_RIGHT_WALL_LANDING_GEAR_CIRCUIT_BREAKER_TOGGLE`
> 
> I can remove if it is unnecessary or confusing

Gotcha, no if you think that makes more sense I'm fine with it - though I'd suggest `PLT_CB_LW_...` or whatever since at least to me it makes more sense to put the common things first, but not a big deal either way

**Comment by acstott** [comment]
> > > I see some of the ids have LW or RW after them - what did that stand for?
> > 
> > 
> > Left Wall / Right Wall - I was trying to retain the nomenclature from DCS as they write out like:
> > `PILOT_RIGHT_WALL_LANDING_GEAR_CIRCUIT_BREAKER_TOGGLE`
> > I can remove if it is unnecessary or confusing
> 
> Gotcha, no if you think that makes more sense I'm fine with it - though I'd suggest `PLT_CB_LW_...` or whatever since at least to me it makes more sense to put the common things first, but not a big deal either way

I reviewed what you had already implemented in the file and there are other left and right wall panel elements without the RW / LW annotation, so I think it is best to leave it off for consistency. Just pushed an update where they are removed

**Comment by charliefoxtwo** [comment]
Is there some sort of line ending disagreement with the auto-generated files? Seems like the entirety of the files were rewritten (including an F-15 json) which isn't ideal 🤔

**Comment by charliefoxtwo** [comment]
@acstott were you able to figure out what's going on with the generated files? I dropped the json files from the repo so that shouldn't be an issue, but Addresses.h still has like 60k changes which isn't ideal.

I think that's the last thing we need to get this merged right? We're nearing the end of the F-4E controls 🎉

**Comment by acstott** [comment]
> @acstott were you able to figure out what's going on with the generated files? I dropped the json files from the repo so that shouldn't be an issue, but Addresses.h still has like 60k changes which isn't ideal.
> 
> I think that's the last thing we need to get this merged right? We're nearing the end of the F-4E controls 🎉

Everything looked good on my end when I tested with Bort, but I didn't do any direct integration tests with actual toggle switches. Addresses.h needs to be regenerated from master with these changes added in from the feature branch, so if things have changed since then (looks like 70+ commits, so lots of changes) maybe the easiest thing would be to kill this PR and open another  to test the file generation?

**Comment by charliefoxtwo** [comment]
> Everything looked good on my end when I tested with Bort, but I didn't do any direct integration tests with actual toggle switches.

Testing with bort is sufficient, no need to actually build a panel or anything :)

> Addresses.h needs to be regenerated from master with these changes added in from the feature branch, so if things have changed since then (looks like 70+ commits, so lots of changes) maybe the easiest thing would be to kill this PR and open another to test the file generation?

Ultimately it's up to you, but you should be able to continue using this PR should you wish - just merge master in, ignore any conflicts in Addresses.h, delete the file, and regenerate it - then you should be good to go! It also makes it easier to review and ensure that only changes to Addresses.h were added. If you'd prefer to rebase or cherry-pick to a new branch though, that's fine too.

**Comment by charliefoxtwo** [comment]
@acstott All other F-4E changes are in for the latest DCS update and the next release of dcs-bios, so the Addresses.h file won't be changing anymore until after circuit breakers are merged (barring any unforeseen bugs).

This is the last feature needed for the next dcs-bios release - do you think you'll be able to have the PR ready this week? If you're busy it's no problem - I'm happy to cherry-pick your changes and resolve the merge conflict myself if it's alright with you.

**Comment by acstott** [comment]
I'll be able to clean it up and cherry pick it tomorrow, can you please give it a once over with Bort just to make sure I didn't miss anything before merging? After that, if there are other open issues you need help with that I can pick up, let me know - I can take a look at open issues but you will know which ones might be easier to tackle than I will at this point

**Comment by charliefoxtwo** [comment]
Sure, happy to test!

This will be the last big item for the F-4. If you're interested in helping out with other aircraft (which is greatly appreciated!), the Kiowa definitely has a lot left. I structured that one a little differently - by panel, rather than device - but otherwise it should be similar. I'll go through and mark some of the simpler panels as [good first issue](https://github.com/DCS-Skunkworks/dcs-bios/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22) in case you own the aircraft and it interests you.

**Comment by acstott** [comment]
Closing in favour of cherry picking + fast-forwarding to open new PR from DCS

**Pull request closed without merging**